PyBASC/utils.py

```diff
@@ -486,295 +409,258 @@ def cluster_matrix_average(M, cluster_assignments):
-   
-   Returns
-    -------
-    A voxel-wise representation of the stabilty matrix.
-   """
-       
-   import random
-   import pandas as pd
-   import numpy as np
-   import time
-   print('debug expand ism1')
-   #import pdb; pdb.set_trace()
-   voxel_num=len(Y1_labels)
-   voxel_ism = np.zeros((voxel_num,voxel_num))
-   transform_mat=np.zeros((len(ism),voxel_num))
- 
-   matrixtime = time.time()
-   print('debug expand ism2') 
-   #import pdb; pdb.set_trace()
-
-   for i in range(0,voxel_num):
-     transform_mat[Y1_labels[i],i]=1
-
-   print('debug expand ism3') 
-
-   temp=np.dot(ism,transform_mat)
-   print('debug expand ism4') 
-
-   target_mat=np.dot(temp.T,transform_mat)
+    ism : individual stabilty matrix. A symmetric array
     
+    Y1_labels : 1-D array of voxel to supervoxel labels, 
+                created in initial data compression
     
-   XM_time= time.time() - matrixtime
-   #print('Matrix expansion took', (time.time() - matrixtime), ' seconds')
-   voxel_ism=target_mat
-            
-   return voxel_ism
+    Returns
+    -------
+    A voxel-wise representation of the stabilty matrix.
+
+    """
+
+    import numpy as np
 
+    voxel_num = len(Y1_labels)
+    voxel_ism = np.zeros((voxel_num, voxel_num))
 
-def data_compression(fmri_masked, mask_img, mask_np, output_size):
+    transform_mat = np.zeros((len(ism), voxel_num))
+    for i in range(voxel_num):
+        transform_mat[Y1_labels[i], i] = 1
+
+    temp = np.dot(ism, transform_mat)
+    voxel_ism = np.dot(temp.T, transform_mat)
+
+    return voxel_ism
+
+
+def data_compression(fmri_masked, mask_img, mask_np, compression_dim):
+    # TODO @AKI update doc
     """
-    data : array_like
-         A matrix of shape (`V`, `N`) with `V` voxels `N` timepoints
-         The functional dataset that needs to be reduced
+    Perform...
+    
+    Parameters
+    ----------
+    data : np.ndarray[ndim=2]
+           A matrix of shape (`V`, `N`) with `V` voxels `N` timepoints
+           The functional dataset that needs to be reduced
     mask : a numpy array of the mask
-    output_size : integer
+    compression_dim : integer
         The number of elements that the data should be reduced to
-        
-    """
-    
-
-## Transform nifti files to a data matrix with the NiftiMasker
-    import time
-    from nilearn import input_data
 
-    datacompressiontime=time.time()
-    nifti_masker = input_data.NiftiMasker(mask_img= mask_img, memory='nilearn_cache',
-                                          mask_strategy='background', memory_level=1,
-                                          standardize=False)
+    Returns
+    -------
+    A dictionaty ...
 
-    ward=[]
+    """
 
-# Perform Ward clustering
     from sklearn.feature_extraction import image
+    from sklearn.cluster import FeatureAgglomeration
+
+    # Perform Ward clustering
     shape = mask_np.shape
     connectivity = image.grid_to_graph(n_x=shape[0], n_y=shape[1],
```

Comment:
Maybe worth playing around with this to figure out how to use the ward.labels_ to transform the connectivity (spatial constraint information) to the supervoxel space, this would help us keep spatial constraints at the supervoxel level and likely boost replicability of clustering even further.